File: sessions/week-01-two-pointers/algorithms/javascript/detectCycleInLinkedList.js
class Node {
    constructor(value) {
        this.value = value;
        this.next = null;
    }
}

function hasCycle(head) {
    let slow = head;
    let fast = head;
    while (fast !== null && fast.next !== null) {
        slow = slow.next;
        fast = fast.next.next;
        if (slow === fast) return true;
    }
    return false;
}

// Example usage
function createCycleList() {
    const head = new Node(1);
    const second = new Node(2);
    const third = new Node(3);
    const fourth = new Node(4);
    head.next = second;
    second.next = third;
    third.next = fourth;
    fourth.next = second;  // cycle here
    return head;
}

const head = createCycleList();
console.log("Cycle detected?", hasCycle(head));

File: sessions/week-01-two-pointers/algorithms/python/detect_cycle_in_linked_list.py
class Node:
    def __init__(self, value):
        self.value = value
        self.next = None

def has_cycle(head):
    slow = fast = head
    while fast and fast.next:
        slow = slow.next
        fast = fast.next.next
        if slow == fast:
            return True
    return False

# Example usage
def create_cycle_list():
    head = Node(1)
    second = Node(2)
    third = Node(3)
    fourth = Node(4)
    head.next = second
    second.next = third
    third.next = fourth
    fourth.next = second  # cycle here
    return head

head = create_cycle_list()
print("Cycle detected?" , has_cycle(head))
